packages/@o3r/rules-engine/src/engine/ruleset-executor.ts

```diff
@@ -195,38 +195,62 @@ export class RulesetExecutor {
     throw new Error(`Unknown condition block met : ${JSON.stringify(nestedCondition)}`);
   }
 
+  // TODO: make public and move to top of the class in a separate PR to make the diff more meaningful in this one
   /**
    * Plug ruleset to fact streams and trigger a first evaluation
-   *
-   * @param ruleset
    */
   public prepareRuleset() {
-    const rulesWithContext: Rule[] = [];
-    const rulesWithoutContext: Rule[] = [];
+    const inputFactsForRule: Record<string, string[]> = {};
+    this.ruleset.rules.forEach((rule) => {
+      const findRuleInputFacts = (obj: AllBlock): string[] => {
+        const ruleInputFacts: string[] = [];
+        // Recursive function to explore each object
+        const exploreObject = (currentObj: any) => {
+          if (currentObj && currentObj.type === 'FACT') {
+            ruleInputFacts.push(currentObj.value);
+          } else if (Array.isArray(currentObj)) {
+            currentObj.forEach((elem: any) => exploreObject(elem));
+          } else {
+            for (const key in currentObj) {
+              if (key === 'operator') {
+                const op = this.operators[currentObj[key]];
+                if (op && op.dependencies) {
+                  ruleInputFacts.push(...op.dependencies);
+                }
+              } else if (typeof currentObj[key] === 'object' && currentObj[key] !== null) {
+                exploreObject(currentObj[key]);
+              }
+            }
+          }
+        };
+        exploreObject(obj);
+        return Array.from(new Set(ruleInputFacts));
+      };
+
+      inputFactsForRule[rule.id] = findRuleInputFacts(rule.rootElement);
```

Comment:
it seems weird to use an Array to add all the items and remove the duplicates at the end
It would make more sense to add the items directly in the Set and convert it to Array at the end:
```suggestion
      const findRuleInputFacts = (obj: AllBlock): string[] => {
        const ruleInputFacts = new Set<string>();
        // Recursive function to explore each object
        const exploreObject = (currentObj: any) => {
          if (currentObj && currentObj.type === 'FACT') {
            ruleInputFacts.add(currentObj.value);
          } else if (Array.isArray(currentObj)) {
            currentObj.forEach(exploreObject);
          } else {
            for (const key in currentObj) {
              if (key === 'operator') {
                const op = this.operators[currentObj[key]];
                if (op && op.dependencies) {
                  op.dependencies.forEach(rulesInputFacts.add, rulesInputFacts);
                }
              } else if (typeof currentObj[key] === 'object' && currentObj[key] !== null) {
                exploreObject(currentObj[key]);
              }
            }
          }
        };
        exploreObject(obj);
        return Array.from(ruleInputFacts);
      };

      inputFactsForRule[rule.id] = findRuleInputFacts(rule.rootElement);
```